File: src/solutions/2023/07/java/Main.java
package kival.aoc.year2023.day07;

import java.util.stream.*;
import java.util.Arrays;
import java.util.Map;
import java.util.HashMap;
import java.util.Collections;

import kival.aoc.core.*;

public class Main {
   private Map<Character, Integer> Ranking = new HashMap<Character, Integer>() {
      {
         put('A', 13);
         put('K', 12);
         put('Q', 11);
         put('J', 10);
         put('T', 9);
         put('9', 8);
         put('8', 7);
         put('7', 6);
         put('6', 5);
         put('5', 4);
         put('4', 3);
         put('3', 2);
         put('2', 1);
         put('1', 0);
      }
   };

   private enum Type {
      HIGH_CARD,
      ONE_PAIR,
      TWO_PAIR,
      THREE_OF_A_KIND,
      FULL_HOUSE,
      FOUR_OF_A_KIND,
      FIVE_OF_A_KIND;
   }

   private record Pair(String v1, Integer v2) {
   }

   private int sortCard(Pair a, Pair b) {
      for (int i = 0; i < 5; i++)
         if (a.v1.charAt(i) != b.v1.charAt(i))
            return Ranking.get(a.v1.charAt(i)) - Ranking.get(b.v1.charAt(i));
      return 0;
   }

   private static int indexOfIntArray(int[] array, int key) {
      int returnvalue = -1;
      for (int i = 0; i < array.length; ++i) {
          if (key == array[i]) {
              returnvalue = i;
              break;
          }
      }
      return returnvalue;
   }
  
   private int getType(String str) {
      int[] values = new int[14];
      Arrays.fill(values, 0);
      for (int i = 0; i < str.length(); i++) {
         values[Ranking.get(str.charAt(i))]++;
      }
      if (values[0] > 0) {
         var temp = values[0];
         values[0] = 0;
         int max = Arrays.stream(values).max().getAsInt();
         int idx = indexOfIntArray(values, max);
         values[idx] += temp;
      }

      values = Arrays.stream(values).filter((e) -> e > 0).toArray();
      if (values.length == 1)
         return Type.FIVE_OF_A_KIND.ordinal();
      if (values.length == 4)
         return Type.ONE_PAIR.ordinal();
      if (values.length == 5)
         return Type.HIGH_CARD.ordinal();

      int min = 5;
      int max = 0;
      for (var v : values) {
         if (min > v)
            min = v;
         if (max < v)
            max = v;
      }
      if (min == 1) {
         if (max == 2)
            return Type.TWO_PAIR.ordinal();
         if (max == 3)
            return Type.THREE_OF_A_KIND.ordinal();
         if (max == 4)
            return Type.FOUR_OF_A_KIND.ordinal();
      }
      return Type.FULL_HOUSE.ordinal();
   }

   private Pair[][] parseInput(String input, boolean joker) {
      return input
            .lines()
            .map(str -> {
               String[] temp = str.split(" ");
               if (joker)
                  temp[0] = temp[0].replaceAll("J", "1");
               return new Pair(temp[0], Integer.parseInt(temp[1]));
            })
            .collect(Collectors.groupingBy(p -> getType(p.v1())))
            .values()
            .stream()
            .map(p -> p.toArray(Pair[]::new))
            .toArray(Pair[][]::new);
   }

   private String solve(String input, boolean joker) {
      var td = parseInput(input, joker);
      int res = 0;
      int i = 0;
      for (var p : td) {
         for (var q : Stream.of(p).sorted(this::sortCard).toArray(Pair[]::new)) {
            res += (Integer)q.v2 * (i + 1);
            i++;
         }
      }
      return String.valueOf(res);
   }

   public String part1(String input, boolean isTest) {
      return solve(input, false);
   }

   public String part2(String input, boolean isTest) {
      return solve(input, true);
   }

   public void main(String[] args) {
      Options.SolutionOptions options = new Options.SolutionOptions();
      Runner.run(args, this::part1, this::part2, options);
   }
}

File: src/solutions/2023/07/csharp/Main.cs

using Core;

namespace Year2023;
public class Day07
{
   static SolutionOptions OPTIONS = new(
   )
   {
      HasAlternate = false,
      HasIO = false,
   };

   private static Dictionary<char, int> Ranking = new Dictionary<char, int>{
         {'A', 13},
         {'K', 12},
         {'Q', 11},
         {'J', 10},
         {'T', 9},
         {'9', 8},
         {'8', 7},
         {'7', 6},
         {'6', 5},
         {'5', 4},
         {'4', 3},
         {'3', 2},
         {'2', 1},
         {'1', 0},
      };

   enum Type : int
   {
      HIGH_CARD,
      ONE_PAIR,
      TWO_PAIR,
      THREE_OF_A_KIND,
      FULL_HOUSE,
      FOUR_OF_A_KIND,
      FIVE_OF_A_KIND,
   }

   private static int SortCard((string, int) a, (string, int) b)
   {
      for (int i = 0; i < 5; i++)
         if (a.Item1[i] != b.Item1[i]) return Ranking[a.Item1[i]] - Ranking[b.Item1[i]];
      return 0;
   }

   private static int GetType(string str)
   {
      int[] values = new int[14].Select((e) => 0).ToArray();
      foreach (var k in str) values[Ranking[k]]++;
      if (values[0] > 0)
      {
         var temp = values[0];
         values[0] = 0;
         var idx = Array.IndexOf(values, values.Max());
         values[idx] += temp;
      }

      values = values.Where((e) => e > 0).ToArray();
      if (values.Length == 1) return (int)Type.FIVE_OF_A_KIND;
      if (values.Length == 4) return (int)Type.ONE_PAIR;
      if (values.Length == 5) return (int)Type.HIGH_CARD;

      int min = 5;
      int max = 0;
      foreach (var v in values)
      {
         if (min > v) min = v;
         if (max < v) max = v;
      }
      if (min == 1)
      {
         if (max == 2) return (int)Type.TWO_PAIR;
         if (max == 3) return (int)Type.THREE_OF_A_KIND;
         if (max == 4) return (int)Type.FOUR_OF_A_KIND;
      }
      return (int)Type.FULL_HOUSE;
   }

   private static (string, int)[][] ParseInput(string input, bool joker)
   {
      return input
         .Split('\n')
         .Select((str) =>
         {
            string[] temp = str.Split(' ');
            if (joker) temp[0] = temp[0].Replace('J', '1');
            return (temp[0], int.Parse(temp[1]));
         })
         .GroupBy((set) => GetType(set.Item1)).OrderBy((group) => group.Key).Select((group) => group.ToArray()).ToArray();
   }

   private static string Solve(string input, bool joker)
   {
      var groups = ParseInput(input, joker);

      int res = 0;
      int i = 1;
      foreach (var group in groups)
      {
         foreach (var set in group.OrderBy(x => x, Comparer<(string, int)>.Create(SortCard)))
         {
            res += set.Item2 * i++;
         }
      }
      return res.ToString();
   }

   static string Part1(string input, bool isTest = false)
   {
      return Solve(input, false);
   }

   static string Part2(string input, bool isTest = false)
   {
      return Solve(input, true);
   }

   static void Main(string[] args)
   {
      Runner.Run(args, Part1, Part2, OPTIONS);
   }
}
